python/lsst/ts/atbuilding/csc/building_csc.py

```diff
@@ -0,0 +1,378 @@
+        )
+
+    async def do_startExtractionFan(self, data):
+        """Implement the ``startExtractionFan`` command."""
+        self.assert_enabled()
+        await self.run_command("start_extraction_fan")
+
+    async def do_stopExtractionFan(self, data):
+        """Implement the ``stopExtractionFan`` command."""
+        self.assert_enabled()
+        await self.run_command("stop_extraction_fan")
+
+    async def run_command(self, command: str) -> None:
+        """Sends a command to the RPi. It writes to the TCP port,
+        and then monitors an `asyncio.Queue`. The response is
+        written to the queue by the `listen_for_messages` method.
+        """
+        await asyncio.wait_for(
+            self.client.write_str(command + "\r\n"), timeout=TCP_TIMEOUT
+        )
+
+        # Wait for a response
+        command_name = command.split()[0]
+        response = await asyncio.wait_for(
+            self.response_queue[command_name].get(), timeout=TCP_TIMEOUT
+        )
+        if response["error"] != 0:
+            # If an error code is supplied, log the error and
+            # raise an exception.
+            self.log.error(
+                "Error response received from command: "
+                + command
+                + " --> "
+                + json.dumps(response)
+            )
+            raise salobj.ExpectedError(json.dumps(response))
+
+        return response
+
+    async def listen_for_messages(self):
+        """Receives messages from the RPi. If the message contains an event
+        (command starts with "evt_") or telemetry (command is "telemetry") it
+        passes the message to the appropriate handler. Otherwise, it sends it
+        to the queue for that command, to be handled by the method that
+        called that command.
+        """
+        while self.client.connected:
+            try:
+                # Receive a message and format it as JSON.
+                self.listen_task = asyncio.create_task(self.client.read_str())
+                message = await asyncio.wait_for(
+                    self.listen_task, timeout=SERVER_MESSAGE_TIMEOUT
+                )
+
+                message = message.strip()
+                message_json = json.loads(message)
+                command = message_json["command"]
+
+                if command in self.callbacks:
+                    # The callbacks dictionary maps telemetry and events to
+                    # handler methods.
+                    await self.callbacks[command](message_json)
+                else:
+                    # Response queues provide the response back to the
+                    # command that sent them.
+                    await self.response_queue[command].put(message_json)
+            except asyncio.TimeoutError:
+                self.log.debug("Timeout while waiting for server response.")
+            except asyncio.CancelledError:
+                pass
```

Comment:
Uhm.. These seems a bit dangerous to ignore a `CancelledError`. This will create a situation where there's no way to stop this loop. Consider either removing this here or interrupting the loop.